tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/Barrier.java

```diff
@@ -0,0 +1,43 @@
+/*
+ *******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.fault.tolerance.tck.bulkhead.lifecycle;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+public class Barrier {
```

Comment:
Do you know if TestNG has something akin to JUnit 4 rules? That's what I'd do if this was JUnit 4.

Reply:
So would I. I did look but TestNG doesn't seem to have anything equivalent. You could probably put together something with an `IInvokedMethodListener` which used reflection to find a field in the class but you'd have to check how it interacts with Arquillian.

The Arquillian extension mechanisms would be another option. I think you'd need to make a remote extension which observes before and after events, and then a local extension to make sure your remote extension gets packaged in each application. Again, you might need to check how that interacts with TestNG's lifecycle.

Lastly, we'd need some way of making what's going on clear to whoever is reading the tests.

A further thought: it might make sense to have an arquillian extension which fires CDI events at the start and end of each test method running on the server. Then we could implement any test fixtures as beans which observe those events.

I'm still on the fence a bit as to whether it's worth doing any of this rather than just having an `@AfterMethod` which adds a little more boilerplate in the test class, which is what we've done elsewhere.

Reply:
I'm thinking you not only need to open the barrier, you also need to wait for all the async operations to finish. That's why I did all the `try`/`catch`/`finally` dance. I'll try to think about this some more.

Reply:
Yeah, that may be true. I'll think on it some more too. For this PR, it's fine as-is.